src/librustc/middle/const_eval.rs

```diff
@@ -987,15 +1001,63 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
           } else {
               ty_hint
           };
-          try!(eval_const_expr_partial(tcx, const_expr, item_hint))
+          try!(eval_const_expr_partial(tcx, const_expr, item_hint, fn_args))
       }
+      ast::ExprCall(ref callee, ref args) => {
+          let (
+            decl,
+            unsafety,
+            abi,
+            block,
+          ) = match try!(eval_const_expr_partial(tcx, callee, UncheckedExprNoHint, fn_args)) {
+              Function(did) => if ast_util::is_local(did) {
+                  match tcx.map.find(did.node) {
+                      Some(ast_map::NodeItem(it)) => match it.node {
+                          ast::ItemFn(
+                              ref decl,
+                              unsafety,
+                              _, // no need to check for constness... either check_const
+                                 // already forbids this or we const eval over whatever
+                                 // we want
+                              abi,
+                              _, // ducktype generics? types are funky in const_eval
+                              ref block,
+                          ) => (decl, unsafety, abi, block),
+                          _ => signal!(e, NonConstPath),
+                      },
+                      _ => signal!(e, NonConstPath),
+                  }
+              } else {
+                  signal!(e, NonConstPath)
+              },
+              _ => signal!(e, NonConstPath),
+          };
+          assert_eq!(decl.inputs.len(), args.len());
+          assert_eq!(unsafety, ast::Unsafety::Normal);
+          assert_eq!(abi, abi::Abi::Rust);
+
+          let mut call_args = NodeMap();
+          for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_val = try!(eval_const_expr_partial(
+                  tcx,
+                  arg_expr,
+                  UncheckedExprNoHint,
```

Comment:
I only know about the ast-type, not the actual type-checked type. I could of course check the appropriate tables if the type is known.